source/agora/common/DNS.d

```diff
@@ -1107,6 +1108,40 @@ public struct Domain
         return result;
     }
 
+    /// Tuple for service and protocol labels
+    alias ServiceProtoTuple = Tuple!(const(char[]), "service", const(char[]), "protocol");
+
+    /***************************************************************************
+
+        Get service and protocol labels
+
+        Returns:
+          A tuple containing both service and protocol labels, empty
+          `ServiceProtoTuple` is returned when one of them is missing.
+
+    ***************************************************************************/
+
+    public ServiceProtoTuple serviceAndProtocolLabels () const nothrow @safe
+    {
+        if (this.value[0] != '_')
+            return ServiceProtoTuple.init;
+
+        try
+        {
+            auto splitted = this.value.splitter('.');
+            const service = splitted.front();
+            splitted.popFront();
+            const proto = splitted.front();
+
+            if (proto[0] != '_')
+                return ServiceProtoTuple.init;
+
+            return ServiceProtoTuple(service[1 .. $], proto[1 .. $]);
+        }
+        catch (Exception)
+            return ServiceProtoTuple.init;
```

Comment:
The only thing that could throw is if UTF-8 decoding fails, so I would make it an `assert(0)`, unless I'm missing something ?